cli/tools/check.rs

```diff
@@ -24,13 +27,65 @@ use crate::cache::CacheDBHash;
 use crate::cache::Caches;
 use crate::cache::FastInsecureHasher;
 use crate::cache::TypeCheckCache;
+use crate::extract;
+use crate::factory::CliFactory;
 use crate::graph_util::BuildFastCheckGraphOptions;
 use crate::graph_util::ModuleGraphBuilder;
 use crate::npm::CliNpmResolver;
 use crate::tsc;
 use crate::tsc::Diagnostics;
 use crate::util::path::to_percent_decoded_str;
 
+pub async fn check(
+  flags: Arc<Flags>,
+  check_flags: CheckFlags,
+) -> Result<(), AnyError> {
+  let factory = CliFactory::from_flags(flags);
+
+  let main_graph_container = factory.main_module_graph_container().await?;
+
+  let specifiers =
+    main_graph_container.collect_specifiers(&check_flags.files)?;
+  if specifiers.is_empty() {
+    log::warn!("{} No matching files found.", colors::yellow("Warning"));
+  }
+
+  let specifiers_for_typecheck = if check_flags.doc {
+    let file_fetcher = factory.file_fetcher()?;
+
+    let mut specifiers_for_typecheck = specifiers
+      .iter()
+      .filter(|specifier| {
+        use MediaType::*;
+        matches!(
+          MediaType::from_specifier(specifier),
+          TypeScript | JavaScript | Tsx | Jsx | Mts | Mjs | Cts | Cjs
```

Comment:
Why do we do this filtering here, but not in the non-doc case?

Reply:
The reason is to preserve the existing `deno test --doc` behavior. For instance, using v1.46.3 of Deno, running `deno test --doc cli/tsc/dts/lib.deno.ns.d.ts` succeeds:

```console
$ deno test --doc cli/tsc/dts/lib.deno.ns.d.ts
Check file:///Users/yusuke/Repo/github.com/magurotuna/deno/cli/tsc/dts/lib.deno.ns$d$ts$22-28.ts
Check file:///Users/yusuke/Repo/github.com/magurotuna/deno/cli/tsc/dts/lib.deno.ns$d$ts$64-69.ts
Check file:///Users/yusuke/Repo/github.com/magurotuna/deno/cli/tsc/dts/lib.deno.ns$d$ts$75-79.ts

... omitted ...

Check file:///Users/yusuke/Repo/github.com/magurotuna/deno/cli/tsc/dts/lib.deno.ns$d$ts$6180-6186.ts
Check file:///Users/yusuke/Repo/github.com/magurotuna/deno/cli/tsc/dts/lib.deno.ns$d$ts$6198-6206.ts

ok | 0 passed | 0 failed (0ms)
```


But if we don't do the filtering here, `./target/debug/deno check --doc cli/tsc/dts/lib.deno.ns.d.ts` does fail.

```console
$ ./target/debug/deno check --doc cli/tsc/dts/lib.deno.ns.d.ts
Check file:///Users/yusuke/Repo/github.com/magurotuna/deno/cli/tsc/dts/lib.deno.ns.d.ts
Check file:///Users/yusuke/Repo/github.com/magurotuna/deno/cli/tsc/dts/lib.deno.ns$d$ts$22-28.ts
Check file:///Users/yusuke/Repo/github.com/magurotuna/deno/cli/tsc/dts/lib.deno.ns$d$ts$64-69.ts

... omitted ...

Check file:///Users/yusuke/Repo/github.com/magurotuna/deno/cli/tsc/dts/lib.deno.ns$d$ts$6180-6186.ts
Check file:///Users/yusuke/Repo/github.com/magurotuna/deno/cli/tsc/dts/lib.deno.ns$d$ts$6198-6206.ts
error: TS2304 [ERROR]: Cannot find name 'TextDecoder'.
    const decoder = new TextDecoder("utf-8");
                        ~~~~~~~~~~~
    at file:///Users/yusuke/Repo/github.com/magurotuna/deno/cli/tsc/dts/lib.deno.ns$d$ts$1007-1032.ts:16:25

TS2304 [ERROR]: Cannot find name 'TextDecoder'.
    const decoder = new TextDecoder("utf-8");
                        ~~~~~~~~~~~
    at file:///Users/yusuke/Repo/github.com/magurotuna/deno/cli/tsc/dts/lib.deno.ns$d$ts$1045-1064.ts:14:25

... omitted ...

TS2304 [ERROR]: Cannot find name 'BufferSource'.
      destination: BufferSource,
                   ~~~~~~~~~~~~
    at file:///Users/yusuke/Repo/github.com/magurotuna/deno/cli/tsc/dts/lib.deno.ns.d.ts:5877:20

TS2304 [ERROR]: Cannot find name 'URL'.
    filename: string | URL,
                       ~~~
    at file:///Users/yusuke/Repo/github.com/magurotuna/deno/cli/tsc/dts/lib.deno.ns.d.ts:6087:24

Found 269 errors.
```

With the filtering, we can get the same result as the current `deno test --doc`.

Reply:
Why don't we want to type check `.d.ts` files or type check their code snippets?

Also, getting the media type from the specifier alone is not accurate for some remote specifiers without an extension.

Reply:
Ah OK I get it now - Fixed [fbc7b8e](https://github.com/denoland/deno/pull/25220/commits/fbc7b8ed4d0ffe39b6c8b2e96af47aa8ebefef21)

Reply:
Oh I now remember why I added the filter - we have a test case to confirm that `deno check --doc cli/tsc/dts/lib.deno.ns.d.ts` (which previously was `deno test --doc`) passes:

https://github.com/denoland/deno/blob/50d87d3c2c88dbf51a399c9dc23d124c99501352/tests/integration/check_tests.rs#L189-L210

The previous command `deno test --doc cli/tsc.dts/lib.deno.ns.d.ts` was successful because it was checking only the snippets in the doc comments, and did not type check the whole file. But since now we move the ability to perform type-check against doc comments into `check --doc` command, it will run type-check against the declaration part of lib.deno.ns.d.ts, resulting in a bunch of type errors. (see the CI error log at https://github.com/denoland/deno/actions/runs/10849599526/job/30109161558?pr=25220#step:43:5705)

Do you have any idea to resolve this issue? My quick idea is that to make `check --doc` only type-check on code snippets in doc comments, but not sure if this really is a good interface for users.

Reply:
After some thoughts, I decided to add `--doc-only` option to the `check` subcommand, which allows us to perform the type checking only on code snippets in JSDoc or markdown. The rationale of having this new flag separately is `deno check` command doesn't accept markdown files, but we want to type check on code snippets in markdown - which `--doc-only` accomplishes.